receiver/mongodbatlasreceiver/alerts.go

```diff
@@ -71,26 +106,123 @@ func newAlertsReceiver(logger *zap.Logger, cfg AlertConfig, consumer consumer.Lo
+			projectInclusionMap[projectClusterKey(pc, inclusion)] = true
+		}
 	}
 
-	s := &http.Server{
-		TLSConfig: tlsConfig,
-		Handler:   http.HandlerFunc(recv.handleRequest),
+	recv := &alertsReceiver{
+		addr:                     cfg.Endpoint,
+		secret:                   cfg.Secret,
+		tlsSettings:              cfg.TLS,
+		consumer:                 consumer,
+		mode:                     cfg.Mode,
+		projects:                 cfg.Projects,
+		retrySettings:            baseConfig.RetrySettings,
+		publicKey:                baseConfig.PublicKey,
+		privateKey:               baseConfig.PrivateKey,
+		wg:                       &sync.WaitGroup{},
+		pollInterval:             baseConfig.Alerts.PollInterval,
+		doneChan:                 make(chan bool, 1),
+		logger:                   logger,
+		projectClusterInclusions: projectInclusionMap,
+		projectClusterExclusions: projectExclusionMap,
+		id:                       baseConfig.ID(),
+		storageID:                baseConfig.StorageID,
 	}
 
-	recv.server = s
+	if recv.mode == alertModePoll {
+		client, err := internal.NewMongoDBAtlasClient(recv.publicKey, recv.privateKey, recv.retrySettings, recv.logger)
+		if err != nil {
+			return nil, err
+		}
+		recv.client = client
+	} else {
+		s := &http.Server{
+			TLSConfig: tlsConfig,
+			Handler:   http.HandlerFunc(recv.handleRequest),
+		}
+		recv.server = s
+	}
 
 	return recv, nil
 }
 
 func (a alertsReceiver) Start(ctx context.Context, host component.Host) error {
+	if a.mode == alertModePoll {
+		return a.startPolling(ctx, host)
+	}
+	return a.startListening(ctx, host)
+}
+
+func (a alertsReceiver) startPolling(ctx context.Context, host component.Host) error {
+	a.logger.Debug("starting alerts receiver in retrieval mode")
+	storageClient, err := adapter.GetStorageClient(ctx, host, a.storageID, a.id)
+	if err != nil {
+		return fmt.Errorf("failed to set up storage: %w", err)
+	}
+	a.storageClient = &storageClient
+	err = a.syncPersistence(ctx)
+	if err != nil {
+		a.logger.Error("there was an error syncing the receiver with checkpoint", zap.Error(err))
+	}
+
+	t := time.NewTicker(a.pollInterval)
+	a.wg.Add(1)
+	go func() {
+		defer a.wg.Done()
+		for {
+			select {
+			case <-t.C:
+				if err := a.retrieveAndProcessAlerts(ctx); err != nil {
+					a.logger.Error("unable to retrieve alerts", zap.Error(err))
+				}
+			case <-a.doneChan:
+				return
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (a alertsReceiver) retrieveAndProcessAlerts(ctx context.Context) error {
+	var alerts []mongodbatlas.Alert
+	for _, p := range a.projects {
+		project, err := a.client.GetProject(ctx, p.Name)
+		if err != nil {
+			a.logger.Error("error retrieving project "+p.Name+":", zap.Error(err))
+			continue
+		}
+		projectAlerts, err := a.client.GetAlerts(ctx, project.ID)
+		if err != nil {
+			a.logger.Error("unable to get alerts for project", zap.Error(err))
+			continue
+		}
+		filteredAlerts := a.applyFilters(p, projectAlerts)
```

Comment:
I've changed it so that this function now does all the filtering, I think it is more apt now since it handles both cluster and timestamp filtering now.

Reply:
👍 I thought about suggesting this but didn't because then you are parsing timestamps twice for every record. Not ideal, but probably worth it for the sake of separating concerns. It can be optimized later if it proves necessary.